storage/src/main/java/fk/prof/storage/S3AsyncStorage.java

```diff
@@ -80,28 +85,57 @@ public void storeAsync(String path, InputStream content, long length) {
         return CompletableFuture.supplyAsync(() -> {
             try {
                 return client.getObject(objectPath.bucket, objectPath.fileName).getObjectContent();
-            }
-            catch (AmazonServiceException svcEx) {
+            } catch (AmazonServiceException svcEx) {
                 LOGGER.error("S3 getObject failed: {}", path, svcEx);
                 throw mapServiceException(svcEx);
-            }
-            catch (AmazonClientException clientEx) {
+            } catch (AmazonClientException clientEx) {
                 LOGGER.error("S3 getObject failed: {}", path, clientEx);
                 throw mapClientException(clientEx);
             }
         }, executorService);
     }
 
+    @Override
+    public CompletableFuture<Set<String>> listAsync(String prefixPath, boolean recursive) {
+        S3ObjectPath objectPath = new S3ObjectPath(prefixPath);
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                Set<String> allObjects = new HashSet<>();
+                ObjectListing objects;
+                if (recursive) {
+                    objects = client.listObjects(new ListObjectsRequest().withBucketName(objectPath.bucket).withPrefix(objectPath.fileName));
+                } else {
+                    objects = client.listObjects(new ListObjectsRequest().withBucketName(objectPath.bucket).withDelimiter(DELIMITER).withPrefix(objectPath.fileName));
+                }
+
+                do {
+                    for (S3ObjectSummary objSummary : objects.getObjectSummaries()) {
+                        allObjects.add(objSummary.getKey());                    //All files with prefix with complete path
```

Comment:
Yes, it is empty when `recursive == false `in our case. This is because when `recursive == false`, `listObjects` is called with both the prefix and the delimiter. This, in our case, where there is no file in the current dir (as it only considers suffixes without the delimiter as files) returns an empty list. It is the `commonPrefixes` that has the distinct folder names (appIds, clusterIds ...) in the current directory .